### PR TITLE
enable deep linking for linux

### PR DESCRIPTION
**Pull request by dunkla**
Before submitting, please confirm you've
 - [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
pull request for
Linux: Support protocol deep linking #627

**Issue link**
Linux: Support protocol deep linking #627

**Test Cases**

**Additional Notes**

### CODE REVIEW

**Comment by wget** [comment]
@dunkla @jasonblais Do we have a scheme like `mattermost://`? IS this a feature we support at all?
Just to understand things properly.
I'll test the commit by building the app on my linux box to see if this change anything in practice.

**Comment by jasonblais** [comment]
Thanks @wget! 

Yes, Mac and Windows support `mattermost://` links - when used, it opens the Desktop App instead of a browser window

Related PRs:
- https://github.com/mattermost/desktop/pull/616
- https://github.com/mattermost/desktop/pull/680
- https://github.com/mattermost/desktop/pull/785

**Comment by yuya-oc** [comment]
Hi @dunkla, thank you for submitting the PR!

I tried this but the deep linking didn't work for me.

Environment:
- Ubuntu 18.04

Steps:
1. Extract .tar.gz
2. Execute `./create_desktop_file.sh` then copy `Mattermost.desktop` to `~/.local/share/applications/`.
3. Open the desktop app.
4. Enter the URL in Chromium: mattermost://community.mattermost.com/core/channels/desktop-app
5. The app should be activated, but that didn't happen.
6. xdg-open also didn't perform.

Are there any extra steps to use deep linking?

**Comment by wget** [comment]
@yuya-oc The patch is working properly, the issue seems to come from the electron app:

```
(master)[wget@hermes ~]$ xdg-open mattermost://community.mattermost.com/core/channels/desktop-app
(process:4556): Gtk-WARNING **: 18:40:55.597: Locale not supported by C library.
        Using the fallback 'C' locale.
(electron) 'app.makeSingleInstance(cb)' is deprecated. Use 'app.requestSingleInstanceLock() and app.on('second-instance', cb)' instead.
In development mode, deeplinking is disabled
Added Extension:  React Developer Tools
/usr/bin/mattermost-desktop: line 3:  4556 Segmentation fault      (core dumped) electron /usr/lib/mattermost-desktop/app.asar --disable-dev-mode "$@"
```

**Comment by yuya-oc** [comment]
@wget The log is weird. `app.makeSingleInstance()` is no longer used in the master branch. And the deep linking should be disabled in development build even when `--disable-dev-mode` is used. So I guess that a wrong executable (probably your local repository's one) was launched by xdg-open.

**Comment by hanzei** [comment]
@dunkla Do you have questions about the [above provided feedback](https://github.com/mattermost/desktop/pull/937#issuecomment-468295209)?

**Comment by dunkla** [comment]
I didn't know you still wanted feedback from me. I thought @yuya-oc said it all.

Nevertheless, i tested it again (fresh build with above patch) and it works for me.

**Comment by esethna** [comment]
@yuya-oc any other feedback on this PR?

**Comment by hanzei** [comment]
@dunkla Sorry for the delay. Would you please merge `master` intro this PR? There has been a lot of changes recently that might have also affected this PR. After that we can test again.

**Comment by dunkla** [comment]
I merged master into the pr. 
`npm test` was fine, the built packages work as expected.

**Comment by hanzei** [comment]
Thanks @dunkla!

@wget @yuya-oc Please review, when you have the time

**Comment by wget** [comment]
@hanzei @dunkla Hello everyone. Sorry for the delay.

The xdg open and mattermost:// links are now working fine from my side. **HOWEVER** the channel change doesn't happen automatically.

If I'm in the default channel Town-Square on the Mattermost Community instance for example, and I open [mattermost://community.mattermost.com/core/channels/desktop-app](mattermost://community.mattermost.com/core/channels/desktop-app), the app is being launched, but we are not being switched to the right Mattermost instance and the right channel.

If this is not intended (a bug), it will need to get fixed first before merging.

**Comment by wget** [comment]
> @dunkla Sorry for the delay. Would you please merge `master` intro this PR? There has been a lot of changes recently that might have also affected this PR. After that we can test again.

And for the record, it's better to rebase rather than merging for this use case :)

**Comment by dunkla** [comment]
@wget I'm lost there, because i cannot reproduce your problem.

Anybody else here, that can confirm/reproduce the problem?

**Comment by hanzei** [comment]
@deanwhillier Can you help with debugging this?

**Comment by jasonblais** [comment]
@yuya-oc or @deanwhillier wondering if you have time to provide troubleshooting/debuggin assistance?

**Comment by deanwhillier** [comment]
I'll give it a try locally and see what I get!

**Comment by wget** [comment]
Checked with the latest 4.3.0-rc2 and the problem persists:

Repro steps:
Create an html document with this:
```
<html>
  <body>
    <a href="mattermost://community.mattermost.com/core/channels/desktop-app">Desktop app deep linking test</a>
  </body>
</html>
```
Open Mattermost Desktop in a channel other than the Desktop app channel.

Click the link from the HTML file. While Mattermost desktop opens, you are not being redirected to the Desktop app channel.

cc @svelle

**Comment by hunterlester** [comment]
I'm not able to reproduce bug where desktop app does open but does not push state to update channel.
Seeing success when running repro steps as recommended above:
Screenshot from 2020-01-11 22-57-17

Think about reasons why others see it fail though...

The `handleDeepLink` method can only be called if it's associated `webview` ref exists in `MainPage` and I have seen instances where the opposite is the case.
In that case the following, which handles pushing the state to expected channel, would not be called:
```
    webview.executeJavaScript(
      'history.pushState(null, null, "' + relativeUrl + '");'
    );
    webview.executeJavaScript(
      'dispatchEvent(new PopStateEvent("popstate", null));'
    );
```

Unfortunately the app is only setup to register the scheme when the package is installed and not when running electron via CLI, so debugging will be a bit tedious.
I'm either going  to spend some time setting up app to register `mattermost://` protocol when running electron through CLI, which is possible, or I'm going to target suspected culprit with well-placed logs.

For example,
_line 298 MainPage.jsx_
```
self.refs[`mattermostView${parsedDeeplink.teamIndex}`].handleDeepLink(parsedDeeplink.path);`
```
assumed that `self.refs[`mattermostView${parsedDeeplink.teamIndex}`].` is true, but I've observed it to not always be the case.

**Comment by hunterlester** [comment]
> Checked with the latest 4.3.0-rc2 and the problem persists:

@wget So I can try to repro more closely, are you saying that you pulled the repo at tag `4.3.0-rc2` and pulled in dunkla's commits?

**Comment by wget** [comment]
> > Checked with the latest 4.3.0-rc2 and the problem persists:
> 
> @wget So I can try to repro more closely, are you saying that you pulled the repo at tag `4.3.0-rc2` and pulled in dunkla's commits?

This is what I did back in the days. Actually I haven't used the git repo directly but rather have modified [the version I maintain on Arch Linux](https://aur.archlinux.org/packages/mattermost-desktop/) and have applied the patch from Dunkla manually. But the effect is the same.

**Comment by hunterlester** [comment]
> This is what I did back in the days. Actually I haven't used the git repo directly but rather have modified [the version I maintain on Arch Linux](https://aur.archlinux.org/packages/mattermost-desktop/) and have applied the patch from Dunkla manually. But the effect is the same.

@wget How did you apply the patch exactly?

Since your app is indeed opening, however not navigating to specific channel, I have a hunch that you've only modified `mattermost.desktop` in the Arch source to handle `mattermost://` protocol and not added the logic for the app to parse the deep link:
```
  if (process.platform === 'win32' || process.platform === 'linux') {
    if (Array.isArray(secondArgv.slice(1)) && secondArgv.slice(1).length > 0) {
      setDeeplinkingUrl(secondArgv.slice(1)[0]);
      mainWindow.webContents.send('protocol-deeplink', deeplinkingUrl);
    }
  }
```

`PKGBUILD` just pulls `https://github.com/mattermost/desktop` from master branch, yes?

**Comment by deanwhillier** [comment]
@Willyfrog, when you have a minute, would you mind doing a review on this please?

**Comment by wget** [comment]
> > This is what I did back in the days. Actually I haven't used the git repo directly but rather have modified [the version I maintain on Arch Linux](https://aur.archlinux.org/packages/mattermost-desktop/) and have applied the patch from Dunkla manually. But the effect is the same.
> 
> @wget How did you apply the patch exactly?
> 
> Since your app is indeed opening, however not navigating to specific channel, I have a hunch that you've only modified `mattermost.desktop` in the Arch source to handle `mattermost://` protocol and not added the logic for the app to parse the deep link:
> 
> ```
>   if (process.platform === 'win32' || process.platform === 'linux') {
>     if (Array.isArray(secondArgv.slice(1)) && secondArgv.slice(1).length > 0) {
>       setDeeplinkingUrl(secondArgv.slice(1)[0]);
>       mainWindow.webContents.send('protocol-deeplink', deeplinkingUrl);
>     }
>   }
> ```
> 
> `PKGBUILD` just pulls `https://github.com/mattermost/desktop` from master branch, yes?

Nope. It just brings the app from the tar.gz published as GitHub releases.

I'm not saying the version that is *published* on Arch Linux has this patch, I'm saying I tested the patch locally (writing a .patch with diff) patching the build process of that version :)

I'll retest everything over the week-end if you want me to :)

File: resources/linux/create_desktop_file.sh [hunk 1]
```diff
@@ -14,5 +14,6 @@ Terminal=false
 Type=Application
 Icon=${FULL_PATH}/icon.svg
 Categories=Network;InstantMessaging;
+MimeType=x-scheme-handler/mattermost;
```
Comment (Willyfrog) [inline resources/linux/create_desktop_file.sh]: people can modify the scheme in [this file](https://github.com/mattermost/desktop/blob/master/electron-builder.json#L24), this script should read from there by using `jq`

**Comment by esethna** [comment]
@deanwhillier @Willyfrog what are the next steps on this PR?

**Comment by Willyfrog** [comment]
@esethna I added a review from @devinbinnie as he has worked on deep linking, once we have all the changes needed to include it, we should decide on next steps, as I'm not sure the original author has seen the activity on this PR.

**Comment by Willyfrog** [comment]
I'm closing this PR as most of the code base has changed, but I'll gladly reopen if it can be reused somehow

**Pull request closed without merging**